src/rt-app_taskgroups.c

```diff
@@ -0,0 +1,365 @@
+next:
+		dir[i] = sep;
+	} while (dir[i]);
+error:
+	free(dir);
+	free(path);
+	return ret;
+}
+
+static int cgroup_rmdir(char *name, int offset)
+{
+	char *dir = NULL, *path = NULL, *sep;
+	int ret = 0;
+
+	if (offset < 0)
+		goto error;
+
+	dir = strdup(name);
+	if (!dir) {
+		log_error("cannot duplicate string [%s]", name);
+		ret = -1;
+		goto error;
+	}
+
+	path = malloc(FILENAME_MAX);
+	if (!path) {
+		perror("malloc");
+		ret = -1;
+		goto error;
+	}
+
+	while (1) {
+		snprintf(path, FILENAME_MAX, "%s%s", tg_ctrl.mount_point, dir);
+
+		ret = rmdir(path);
+		if (ret) {
+			switch (errno) {
+			case ENOTEMPTY:
+				log_debug("cgroup [%s] not empty, continue ...", dir);
+				ret = 0;
+				break;
+			case EBUSY:
+				log_debug("cgroup [%s] is busy, continue ...", dir);
+				ret = 0;
+				break;
+			case ENOENT:
+				log_debug("cgroup [%s] doesn't exist, continue ...", dir);
+				ret = 0;
+				break;
+			default:
+				log_error("cgroup [%s] unhandled error (%s)", dir, strerror(errno));
+				break;
+			}
+			break;
+		}
+
+		log_debug("cgroup [%s] removed", dir);
+
+		sep = strrchr(dir, '/');
+		if (!sep || sep <= dir + offset)
+			break;
+		*sep = '\0';
+	}
+error:
+	free(dir);
+	free(path);
+	return ret;
+}
+
+static void create_taskgroup(taskgroup_t *tg)
+{
+	if (!tg)
+		return;
+
+	log_debug("create taskgroup [%s]", tg->name);
+
+	if (cgroup_mkdir(tg->name, &tg->offset)) {
+		log_critical("cannot create taskgroup [%s]", tg->name);
+		exit(EXIT_FAILURE);
+	}
+}
+
+void create_taskgroups(rtapp_options_t *opts)
+{
+	int i, y;
+
+	for (i = 0; i < opts->nthreads; i++) {
+		thread_data_t *tdata = &opts->threads_data[i];
+
+		for (y = 0; y < tdata->nphases; y++) {
+			phase_data_t *pdata = &tdata->phases[y];
+
+			create_taskgroup(pdata->taskgroup);
+		}
+	}
+}
+
+static void remove_taskgroup(taskgroup_t *tg)
+{
+	if (!tg)
+		return;
+
+	log_debug("remove taskgroup [%s]", tg->name);
+
+	if (cgroup_rmdir(tg->name, tg->offset)) {
+		log_critical("cannot remove taskgroup [%s]", tg->name);
+		exit(EXIT_FAILURE);
+	}
+}
+
+void remove_taskgroups(rtapp_options_t *opts)
+{
+	int i, y;
+
+	/*
+	 * Remove taskgroups in the reverse order they were created in to let
+	 * the taskgroup->offset ownership mechanism work.
+	 */
+	for (i = opts->nthreads - 1; i >= 0; i--) {
+		thread_data_t *tdata = &opts->threads_data[i];
+
+		for (y = tdata->nphases - 1; y >= 0; y--) {
+			phase_data_t *pdata = &tdata->phases[y];
+
+			remove_taskgroup(pdata->taskgroup);
+		}
+	}
+}
+
+void set_taskgroup(thread_data_t *data, taskgroup_t *tg)
+{
+	if (!tg)
+		return;
+
+	if (data->curr_taskgroup && !strcmp(data->curr_taskgroup->name, tg->name))
+		return;
+
+	log_debug("[%d] set task [%s] taskgroup [%s]", data->ind, data->name, tg->name);
+
+	if (cgroup_attach_task(tg->name, data->tid)) {
+		log_critical("cannot attach task to taskgroup [%s]", tg->name);
+		exit(EXIT_FAILURE);
+	}
+
+	data->curr_taskgroup = tg;
+}
+
+void reset_taskgroup(thread_data_t *data)
+{
```

Comment:
why do you need to reset taskgroup of thread data when thread exit

Reply:
Testing has revealed that reset_taskgroup() is indeed needed. Occasionally the thread is not removed from the taskgroup yet so removal of the taskgroup fails. So I add the following patch on top of the stack:

```
    taskgroups: Empty taskgroup before removal
    
    Force thread into root taskgroup at the end of its start routine. This
    makes sure that a taskgroup is empty before it is removed.
    
    Tests have shown that occasionally the taskgroup being removed is not
    empty so cgroup_rmdir() returns EBUSY.
    
    Signed-off-by: Dietmar Eggemann <dietmar.eggemann@arm.com>

diff --git a/src/rt-app.c b/src/rt-app.c
index fd7e90afd576..77e603eb44cd 100644
--- a/src/rt-app.c
+++ b/src/rt-app.c
@@ -1168,6 +1168,9 @@ void *thread_body(void *arg)
                exit(EXIT_FAILURE);
        }
 
+       /* Force thread into root taskgroup. */
+       reset_thread_taskgroup();
+
        if (timings) {
                int j;
 
diff --git a/src/rt-app_taskgroups.c b/src/rt-app_taskgroups.c
index efa2d794555b..4c92c4ff57dd 100644
--- a/src/rt-app_taskgroups.c
+++ b/src/rt-app_taskgroups.c
@@ -93,6 +93,14 @@ void set_thread_taskgroup(thread_data_t *data, taskgroup_data_t *tg)
        data->curr_taskgroup_data = tg;
 }
 
+void reset_thread_taskgroup(void)
+{
+       if (!ctrl.nr_tgs)
+               return;
+
+       cgroup_attach_task("/");
+}
+
 static int cgroup_check_cpu_controller(void)
 {
        int dummy[2], enabled, ret = 0;
diff --git a/src/rt-app_taskgroups.h b/src/rt-app_taskgroups.h
index 64421a8394b0..756284c72cb6 100644
--- a/src/rt-app_taskgroups.h
+++ b/src/rt-app_taskgroups.h
@@ -6,6 +6,7 @@
 taskgroup_data_t *alloc_taskgroup(size_t size);
 taskgroup_data_t *find_taskgroup(char *name);
 void set_thread_taskgroup(thread_data_t *data, taskgroup_data_t *tg);
+void reset_thread_taskgroup(void);
 
 void initialize_cgroups(void);
```